haystack_experimental/components/retrievers/auto_merging_retriever.py

```diff
@@ -117,41 +117,58 @@ def run(self, matched_leaf_documents: List[Document]):
-            if len(parent_doc) == 0:
-                raise ValueError(f"Parent document with id {doc_id} not found in the document store.")
-            if len(parent_doc) > 1:
-                raise ValueError(f"Multiple parent documents found with id {doc_id} in the document store.")
-            if not parent_doc[0].meta.get("__children_ids"):
-                raise ValueError(f"Parent document with id {doc_id} does not have any children.")
-            parent_children_count = len(parent_doc[0].meta["__children_ids"])
-
-            # return either the parent document or the matched leaf documents based on the threshold value
-            score = len(retrieved_child_docs) / parent_children_count
-            if score >= self.threshold:
-                # return the parent document
-                docs_to_return.append(parent_doc[0])
-            else:
-                # return all the matched leaf documents which are child of this parent document
-                leafs_ids = {doc.id for doc in retrieved_child_docs}
-                docs_to_return.extend([doc for doc in matched_leaf_documents if doc.id in leafs_ids])
-
-        return {"documents": docs_to_return}
+        AutoMergingRetriever._check_valid_documents(matched_leaf_documents)
+
+        def try_merge_level(documents: List[Document]) -> List[Document]:
+            if not documents:
+                return []
+
+            docs_to_return = []
+            parent_documents: Dict[str, List[Document]] = defaultdict(list)  # to group the documents by their parent
+
+            for doc in documents:
+                if doc.meta.get("__parent_id"):  # only docs that have parents
+                    parent_documents[doc.meta["__parent_id"]].append(doc)
+                else:
+                    docs_to_return.append(doc)  # keep docs that have no parents
+
+            # If no documents have parents, we're done
+            if not parent_documents:
+                return documents
+
+            # Process each parent group
+            merged_docs = []
+            for doc_id, child_docs in parent_documents.items():
+                parent_doc = self.document_store.filter_documents({"field": "id", "operator": "==", "value": doc_id})
+                if len(parent_doc) != 1:
+                    raise ValueError(f"Expected 1 parent document with id {doc_id}, found {len(parent_doc)}")
+
+                parent = parent_doc[0]
+                if not parent.meta.get("__children_ids"):
+                    raise ValueError(f"Parent document with id {doc_id} does not have any children.")
+
+                # Calculate merge score
+                score = len(child_docs) / len(parent.meta["__children_ids"])
+                if score > self.threshold:
+                    merged_docs.append(parent)  # Merge into parent
+                else:
+                    merged_docs.extend(child_docs)  # Keep children separate
```

Comment:
I think we could extend docs_to_return here instead. So `docs_to_return.extend(child_docs)`

Reply:
this is yet another good improvement, thanks 👍🏽